client/src/components/campaignCreatorComponents/CampaignCreator.tsx

```diff
@@ -36,8 +43,81 @@ const CampaignCreator = ({ campaignSettings, setCampaignSettings }: CampaignCrea
         navigate('/twilight-library/dashboard');
     };
 
+    const addToTimeline = (
+        timeline: Record<number, string[]>,
+        nodeKey: keyof NodePillarLists,
+        typeKey: keyof TimelineOptionList,
+    ) => {
+        if (!selectedCampaign.flexible_nemesis_encounters && typeKey === 'nemesis') {
+            // Handle strict Nemesis timeline
+            const nemesis_tier: string = nodeKey.slice(-1);
+            selectedCampaign[nodeKey].forEach((selection: string) => {
+                [1, 2, 3].forEach((level: number) => {
+                    const fight_str: string = `nn${nemesis_tier}_lvl${level}_fight_year`;
+                    const fight_key: keyof NemesisFightYearLists = fight_str as keyof NemesisFightYearLists;
+                    const fight_year: number | null = selectedCampaign[fight_key];
+                    if (fight_year !== null) {
+                        timeline[fight_year].push(`NE - ${selection} lvl ${level}`);
+                    }
+                });
+            });
+            return;
+        }
+
+        selectedCampaign[nodeKey].forEach((selection: string) => {
+            const query: Record<string, Record<string, string[]>> = campaignOptionsData.timeline[typeKey];
+            if (query.hasOwnProperty(selection)) {
```

Comment:
Two things:
Start using `map` and `filter` in place of `forEach` - map to transform data (which can then be pushed to timeline), filter to conditionally do so.

Also, if this isn't user input data, there's little reason to verify it it's your data own it

Reply:
Basically any time you're using `.push`, ask yourself if you can use `.map`

Reply:
understand. note for the highlighted code, we could not find a clean solution using map. will leave as push. will check other portions of the code however.

Reply:
removed other instances of push outside of the above code section